### PR TITLE
Resource Views Work with All Kubernetes Models

### PR DESCRIPTION
**Pull request by alecmerdler**
### Description

Significant reduction of [tech debt](https://github.com/openshift/console/pull/98#discussion-diff-193873124R90) related to Kubernetes models and resource view components. Maintains backwards-compatibility with `plural`-style URLs, but enables defining two CRDs with the same `kind` or `plural` names and having them viewable in the console.

### Changes
- Resource view components are keyed by their associated model's `GroupVersionKind`
- `connectToPlural` works with runtime and statically-defined models
- `DetailsPage` actually passes extra `props.resources` to its child components

Addresses https://jira.coreos.com/browse/CONSOLE-562 
Fixes https://jira.coreos.com/browse/CONSOLE-572

### CODE REVIEW

File: frontend/public/components/cloud-services/clusterserviceversion-resource.tsx [hunk 1]
```diff
@@ -222,29 +222,23 @@ export const ClusterServiceVersionResourceDetails = connectToPlural(
...
-      {name: 'Resources', href: 'resources', component: (resourcesProps) => <Resources {...resourcesProps} clusterServiceVersion={props.csv.data} />},
-    ]}
-  /> }
-</div>;
-
-export const ClusterServiceVersionResourcesDetailsPage: React.SFC<ClusterServiceVersionResourcesDetailsPageProps> = (props) => (
-  <Firehose resources={[{kind: referenceForModel(ClusterServiceVersionModel), name: props.match.params.appName, namespace: props.namespace, isList: false, prop: 'csv'}]}>
-    <CSVResourceDetails {...props} />
-  </Firehose>);
+export const ClusterServiceVersionResourcesDetailsPage: React.SFC<ClusterServiceVersionResourcesDetailsPageProps> = (props) => <DetailsPage
```
Comment (alecmerdler) [inline frontend/public/components/cloud-services/clusterserviceversion-resource.tsx]: Utilizing multi-detail view component here to reduce a level of components.

File: frontend/public/module/k8s/k8s-reducers.ts [hunk 2]
```diff
@@ -87,8 +87,6 @@ export default (state: ImmutableMap<string, any>, action) => {
     case types.resources:
       return action.resources.models
         .filter(model => !state.getIn(['RESOURCES', 'models']).has(referenceForModel(model)))
-        // FIXME: Need to remove duplicate `kinds` of static models because it breaks `connectToPlural()`. This means a CRD of `kind: Pod` cannot be defined!
-        .filter(model => !state.getIn(['RESOURCES', 'models']).find(({kind}) => kind === model.kind))
```
Comment (spadgett) [inline frontend/public/module/k8s/k8s-reducers.ts]: 👍

File: frontend/public/components/factory/details.tsx [hunk 3]
```diff
@@ -23,7 +23,12 @@ export const DetailsPage: React.SFC<DetailsPageProps> = (props) => <Firehose res
   prop: 'obj',
 } as FirehoseResource].concat(props.resources || [])}>
   <NavTitle detail={true} title={props.name} menuActions={props.menuActions} kind={props.kind} breadcrumbsFor={props.breadcrumbsFor} />
-  <VertNav pages={props.pages} className={`co-m-${_.get(props.kind, 'kind', props.kind)}`} match={props.match} label={props.label || (props.kind as any).label} />
+  <VertNav
+    pages={props.pages}
+    className={`co-m-${_.get(props.kind, 'kind', props.kind)}`}
+    match={props.match}
+    label={props.label || (props.kind as any).label}
+    resourceKeys={_.map(props.resources, ({prop}) => prop)} />
```
Comment (spadgett) [inline frontend/public/components/factory/details.tsx]: Is this the same as

```js
_.map(props.resources, 'prop')
```

File: frontend/public/kinds.ts [hunk 4]
```diff
@@ -17,18 +15,23 @@ export const connectToModel = connect((state: State, props: {kind: K8sResourceKi
...
  */
-export const connectToPlural = connect((state: State, props: {plural?: string, match: match<{plural: string}>}) => {
+export const connectToPlural = connect((state: State, props: {plural?: GroupVersionKind | string, match: match<{plural: GroupVersionKind | string}>}) => {
   const plural = props.plural || _.get(props, 'match.params.plural');
-  const [refKey, kindObj] = state.k8s.getIn(['RESOURCES', 'models']).findEntry((v, k) => v.plural === plural || k === plural) || ['', undefined];
 
-  return {kindObj, modelRef: refKey, kindsInFlight: state.k8s.getIn(['RESOURCES', inFlight])} as any;
+  const kindObj = !isGroupVersionKind(plural)
+    ? _.find(k8sModels, model => model.plural === plural)
+    : state.k8s.getIn(['RESOURCES', 'models']).get(plural);
```
Comment (spadgett) [inline frontend/public/kinds.ts]: nit: I think it's easier to read if you flip the logic here

```js
const kindObj = isGroupVersionKind(plural)
  ? state.k8s.getIn(['RESOURCES', 'models']).get(plural)
  : _.find(k8sModels, model => model.plural === plural);
```

File: frontend/public/kinds.ts [hunk 5]
```diff
@@ -17,18 +15,23 @@ export const connectToModel = connect((state: State, props: {kind: K8sResourceKi
...
+
+  const modelRef = !isGroupVersionKind(plural) ? kindObj.kind : plural;
```
Comment (spadgett) [inline frontend/public/kinds.ts]: Same comment here

File: frontend/public/components/resource-pages.ts [hunk 6]
```diff
@@ -34,57 +36,104 @@ import { HorizontalPodAutoscalersPage, HorizontalPodAutoscalersDetailsPage } fro
...
+  StorageClassModel,
+  CustomResourceDefinitionModel,
+  ClusterServiceVersionModel,
+  SubscriptionModel,
+  InstallPlanModel,
+  ImageStreamTagModel,
+  ClusterRoleModel,
+  ContainerModel,
+  CatalogSourceModel,
+} from '../models';
```
Comment (spadgett) [inline frontend/public/components/resource-pages.ts]: Should we consider `import * as models from '../models';`?

**Comment by spadgett** [comment]
jenkins rebuild

File: frontend/public/components/resource-pages.ts (hunk 6 above)
Comment (alecmerdler) [reply frontend/public/components/resource-pages.ts]: I'd rather be explicit about the models we import (not all of them have resource list/detail views).

File: frontend/public/components/utils/vertnav.jsx [hunk 7]
```diff
@@ -76,18 +76,19 @@ export const NavBar = ({pages, basePath}) => {
...
 
     const componentProps = _.pick(props, ['filters', 'selected', 'match']);
     componentProps.obj = props.obj.data;
+    const extraResources = _.reduce(props.resourceKeys, (acc, key) => ({...acc, [key]: props[key].data}), {});
 
     const routes = props.pages.map(p => {
       const path = `${props.match.url}/${p.href}`;
       const render = () => {
-        return <p.component {...componentProps} />;
+        return <p.component {...componentProps} {...extraResources} />;
```
Comment (alecmerdler) [inline frontend/public/components/utils/vertnav.jsx]: Actually pass the extra resources to the sub-components.

File: frontend/public/module/k8s/k8s-reducers.ts [hunk 8]
```diff
@@ -87,8 +87,10 @@ export default (state: ImmutableMap<string, any>, action) => {
...
+        .filter(model => {
+          const existingModel = state.getIn(['RESOURCES', 'models', model.kind]);
+          return !existingModel || referenceForModel(existingModel) !== referenceForModel(model);
```
Comment (alecmerdler) [inline frontend/public/module/k8s/k8s-reducers.ts]: Doesn't register duplicate models (i.e. `Pods`).

**Pull request merged**